sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,938 @@
+   */
+  public static <K, V> Read<K, V> read() {
+    return new AutoValue_HadoopInputFormatIO_Read.Builder<K, V>().build();
+  }
+
+  /**
+   * A {@link PTransform} that reads from for any data source which implements Hadoop InputFormat.
+   * For e.g. Cassandra, Elasticsearch, HBase, Redis, Postgres etc. See the class-level Javadoc on
+   * {@link HadoopInputFormatIO} for more information.
+   * 
+   * @param <K> Type of keys to be read.
+   * @param <V> Type of values to be read.
+   * @see HadoopInputFormatIO
+   */
+  @AutoValue
+  public abstract static class Read<K, V> extends PTransform<PBegin, PCollection<KV<K, V>>> {
+
+    public static TypeDescriptor<?> inputFormatClass;
+    public static TypeDescriptor<?> inputFormatKeyClass;
+    public static TypeDescriptor<?> inputFormatValueClass;
+
+    /**
+     * Returns the Hadoop Configuration which contains specification of source.
+     */
+    @Nullable public abstract SerializableConfiguration getConfiguration();
+    @Nullable public abstract SimpleFunction<?, K> getKeyTranslationFunction();
+    @Nullable public abstract SimpleFunction<?, V> getValueTranslationFunction();
+    @Nullable public abstract TypeDescriptor<K> getKeyClass();
+    @Nullable public abstract TypeDescriptor<V> getValueClass();
+
+    abstract Builder<K, V> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<K, V> {
+      abstract Builder<K, V> setConfiguration(SerializableConfiguration configuration);
+      abstract Builder<K, V> setKeyTranslationFunction(SimpleFunction<?, K> function);
+      abstract Builder<K, V> setValueTranslationFunction(SimpleFunction<?, V> function);
+      abstract Builder<K, V> setKeyClass(TypeDescriptor<K> keyClass);
+      abstract Builder<K, V> setValueClass(TypeDescriptor<V> valueClass);
+      abstract Read<K, V> build();
+    }
+
+    /**
+     * Returns a new {@link HadoopInputFormatIO.Read} that will read from the source using the
+     * options indicated by the given configuration.
+     *
+     * <p>
+     * Does not modify this object.
+     */
+    public Read<K, V> withConfiguration(Configuration configuration) {
+      validateConfiguration(configuration);
+      inputFormatClass = TypeDescriptor
+          .of(configuration.getClass(HadoopInputFormatIOContants.INPUTFORMAT_CLASSNAME, null));
+      inputFormatKeyClass = TypeDescriptor
+          .of(configuration.getClass(HadoopInputFormatIOContants.KEY_CLASS, null));
+      inputFormatValueClass = TypeDescriptor
+          .of(configuration.getClass(HadoopInputFormatIOContants.VALUE_CLASS, null));
+      // Sets the configuration.
+      Builder<K, V> builder = toBuilder()
+          .setConfiguration(new SerializableConfiguration(configuration));
+      /*
+       * Sets the output key class to InputFormat key class if withKeyTranslation() is not called
+       * yet.
+       */
+      if (this.getKeyClass() == null) {
```

Comment:
This is a good catch, we ran by calling it twice, so current behavior is that the key and value class of first config are set , inputformat class of second config is set. This is a bug. Adding a unit test to cover this scenario and fixing it. Thanks for throwing this idea.